### ISSUE TITLE
PR 29898 breaks the build for hpctoolkit and MPI C++ on Cray

### ISSUE THREAD

**Issue by mwkrentel** [LLNL/spack#30223]
### Steps to reproduce

Commit 6e5cba7b8269 from PR #29898 on Apr 11 breaks the build for
hpctoolkit on cray.  It probably breaks the build for any package
wanting to build an MPI C++ program on cray from inside spack.

On cray, the MPI C++ wrapper is something like `/opt/cray/pe/craype/2.6.5/bin/CC`.
In hpctoolkit configure, we test MPICXX to make sure that it can
compile MPI code and that the code is compatible with CXX.

```
#include <mpi.h>
int myRank(void)
{
  int rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  return rank;
}
```

The `cray-mpich/7.7.14` sets a bunch of environ variables that the CC
wrapper needs.  Without this module, the above snippet fails to compile
with:

```
configure:19823: checking if MPICXX is a valid MPI C++ compiler
configure:19850: /opt/cray/pe/craype/2.6.5/bin/CC -c   conftest.cpp >&5
conftest.cpp:52:10: fatal error: mpi.h: No such file or directory
   52 | #include <mpi.h>
      |          ^~~~~~~
compilation terminated.
```

Of course, it compiles fine from the shell outside of spack.  It also
fails if I explicitly add `cray-mpich/7.7.14` to the list of compiler
modules.  And it even fails with `--dirty`.  That should not happen.
`--dirty` is supposed to keep my environment, not unload modules.

----------

I'm not sure of the rationale for unloading `cray-mpich`.  But I'm
going to need it left loaded to compile MPI C++ code on cray.

I can suggest one adjustment where we can all get along: make the
module unload conditional on not `--dirty`.  That is, dirty overrides
the module unload, which it should anyway, that's what dirty means,
keep my environment intact.  I'm guessing that would be a small, easy
fix.

That does mean that building hpctoolkit for cray will now require
dirty.  Cray is so complicated that I don't know if I consider that
good, bad, hilarious or what.

Ok?  Any better ideas?

@skosukhin @becker33 @alalazo 

### Information on your system

spack develop as of april 19, 2022.

* **Spack:** 0.18.0.dev0 (abc162cf3b39ae1907e4bad3b27cb704feb4f0a8)
* **Python:** 3.6.12
* **Platform:** cray-sles15-haswell
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by haampie** [comment]
> I'm not sure of the rationale for unloading cray-mpich

A series of poor decisions by Cray: (1) their compiler wrappers do overlinking by default, for every loaded module they make pkg-config generate extra linker flags. (2) they provide a libmpi that can't locate its own dependencies, not even when loading the cray-mpich module, you need to manually load its _dependencies_ that populate LD_LIBRARY_PATH to make libmpi happy.

(1) and (2) combined is what's causing issues. Overlinking is usually a no-op, except when the linker can't locate the dependencies of the redundant library it tries to link, then it fails. So you end up in a situation where you build zlib and it fails with a linking error about MPI.

But I don't see how this is an issue for you, since you apparently depend on mpi, and if you add it as an external and provide the module, it will first be unloaded when setting up a clean build env, and then be loaded as part of setting up dependencies?

**Comment by mwkrentel** [comment]
@haampie Also an unfortunate conflict between spack and cray.
The cray MPI C++ wrapper is `CC`, but spack puts CC on my path ahead
of the cray wrapper, and the spack CC is not an MPI compiler.

This caused me to go deeper in my PATH to find the cray CC and use
that.  And that fails without the cray-mpich module.

I'll see if I can make it work with the external for the cray MPI.
That would be more thematic.

But wouldn't you agree that `--dirty` should not unload the module?
If I want all the PrgEnv, etc stuff, then --dirty is how I ask for it.
That has been my workaround for difficult things like this.

**Comment by mwkrentel** [comment]
Ok, I see what happened.  I added current support for cray + mpi to
hpctoolkit in Dec 2018, and the cray-mpich module was added Dec 2020.
I'll try reworking configure to use cray-mpich, that should simplify things.

But I still request that `--dirty` should not unload any modules.
Dirty means "leave my environment intact" and unloading modules
deletes essential variables from my environment.  It's an important
workaround.
